src/sentry/api/endpoints/project_details.py

```diff
@@ -264,9 +264,19 @@ def validate_relayPiiConfig(self, value):
         return validate_pii_config_update(organization, value)
 
     def validate_relayCustomMetricCardinalityLimit(self, value):
-        if value is not None and value < 0:
+        if value is None:
+            return value
+
+        if value < 0:
             raise serializers.ValidationError("Cardinality limit must be a non-negative integer.")
 
+        # Value is stored as uint32 in relay
+        # TODO: find a way to share this constant between relay and sentry
+        if value > 4_000_000_000:
```

Comment:
why 4 million instead of the actual uint32 limit?

Reply:
Mostly for the error message being nice to read. But agree from a code perspective it is just more confusing.
Will change to the uint32 limit 👍